Implement batched query support

InBetweenNames opened the pull request:
* Only Dense queries are accelerated currently
* Certain features are awaiting on GATs/generic const expressions
* "Alignment-oblivious" architecture designed for best-in-class performance on modern (>2009) microarchitectures for simplicity

# Objective
Fixes #1990

## Solution

- Implement a batch-oriented view of the Bevy ECS suitable for SIMD acceleration, providing overalignment of table columns to ensure efficient aligned access where possible.
---

## Changelog

* Queries can be executed in batches of `N` using the `for_each_batched`/`for_each_mut_batched` interfaces
* Added a `WorldQueryBatch`` trait to support batched queries
* Performance related changes for batches:
  * Added a `MAX_SIMD_ALIGNMENT` constant that table columns are aligned to (at minimum) (cache related)
  * `BlobVec`s are now aligned to `MAX_SIMD_ALIGNMENT` to support batching use cases and better performance (aligned loads)
  * `SimdAlignedVec` type was added to provide `MAX_SIMD_ALIGNMENT` on the inner buffer, providing `Vec`-like semantics (aligned loads)
* Tests have been added for the new functionality
* Docs too (with runnable examples)
* Benchmarks provided directly in the Bevy tree (requires AVX to run!  Add `-Ctarget-feature=+avx` to your `RUSTFLAGS` on x86)

A detailed discussion can be found in #1990 , including benchmarks justifying why batched queries are worth implementing.
Older benchmarks can be found here: https://github.com/inbetweennames/bevy-simd-bench

## Migration Guide

No breaking changes.

Fixes #6161

alice-i-cecile commented on crates/bevy_ecs/src/change_detection.rs:
```diff
@@ -271,6 +283,143 @@ change_detection_impl!(Mut<'a, T>, T,);
+    }
+}
+
+//change_detection_impl!(MutBatch<'a, T>, T,);
```
Ah; we impl these below. Just remove the commented code.

alice-i-cecile replied:
Leave a comment about why we can't use the macro though :)

InBetweenNames replied:
Done!

alice-i-cecile commented on crates/bevy_ptr/src/batch.rs:
```diff
@@ -0,0 +1,282 @@
+
+/// This trait allows generic code using [`AlignedBatchT`] with varying sizes and alignments
+/// to call methods on the batches.
+pub trait AlignedBatchTrait<T, const N: usize>: private::SealedAlignedBatchTrait {
```
Maybe `AlignedBatchOps`?

InBetweenNames replied:
Yes agreed, this is a lot clearer for the intent

alice-i-cecile commented on crates/bevy_ptr/src/batch.rs:
```diff
@@ -0,0 +1,282 @@
+//batch lambda accepts arguments of AlignedBatchTrait<T, N> directly. This will let the alignment be automatically computed
+//and allow different query elements to have different alignments.
+
+/// This trait allows generic code using [`AlignedBatchT`] with varying sizes and alignments
```
This needs doc comments explaining why it's a sealed trait, with links to reading material about the pattern.

InBetweenNames replied:
Just a quick question before I add this in -- `bevy_ptr` has a sealed `UnsafeCellDeref` trait in the root, should the documentation on sealed traits be linked there too?

alice-i-cecile replied:
Yes, but make a new issue for that please.

alice-i-cecile commented on crates/bevy_ecs/src/query/fetch.rs:
```diff
@@ -444,6 +448,50 @@ pub trait WorldQueryGats<'world> {
 }
 
+/// An extension of [`WorldQuery`] for batched queries.
+///
```
```suggestion
```

alice-i-cecile commented on crates/bevy_ecs/src/query/fetch.rs:
```diff
@@ -444,6 +448,50 @@ pub trait WorldQueryGats<'world> {
+pub trait WorldQueryBatch<const N: usize, const ALIGN: usize>:
+    for<'w> WorldQueryBatchGats<'w, N, ALIGN> + for<'w> WorldQueryGats<'w>
+{
+    /// Retrieve a batch of size `N` with desired alignment `ALIGN` from the current table.
```
This need a little more detail on how to select (and verify) the correct alignment.

InBetweenNames replied:
Agreed, but I'm not sure where the best place for that documentation should be.  In general the alignment you'd choose would be the greatest alignment possible for your chosen batch.  There's a compile time assertion preventing you from getting it wrong, fortunately :) the expectation is that in the future when "generic const expressions" are stable, you won't need to type this at all anymore since it'll be computed for you (the batch alignment is actually just `gcd(MAX_SIMD_ALIGNMENT, N * size_of::<T>())`).  Having to enter the alignment manually is just a stopgap measure until that's stable.

As an example for how to choose `ALIGN`, consider a batch of 4 `Vec3`s.  A single `Vec3` is 12 bytes and a batch of 4 is 48 bytes.   Since all `BlobVec`s start on `MAX_SIMD_ALIGNMENT` boundaries (64 bytes), this means that every batch will start on a 16 byte boundary guaranteed: the linear combination of `blob_vec_addr + i*48` will always be divisible by 16.  In practice this means we can just check whether `ALIGN` divides `N * size_of::<T>()`.

To make it easy and safe, a compile time assertion will check that `ALIGN` divides `N * size_of::<T>()` and fail the build if it doesn't.  This is the safest option as it would lead to UB otherwise or require runtime checks.

InBetweenNames replied:
I added a little more detail on this in one of my latest commits and placed it in `for_each_mut_batched`.  With the recent decision to wait until GAT stabilization, more detail might not be needed, as it will be automatically derived :) is it OK if I resolve this, in light of that?

BoxyUwU replied:
GATs is being stabilized not generic const exprs

InBetweenNames replied:
In that case it will still be necessary to manually specify `ALIGN`.  I'll see what can be done though.

alice-i-cecile commented on crates/bevy_ecs/src/query/fetch.rs:
```diff
@@ -455,15 +503,21 @@ pub unsafe trait ReadOnlyWorldQuery: WorldQuery<ReadOnly = Self> {}
 pub type QueryFetch<'w, Q> = <Q as WorldQueryGats<'w>>::Fetch;
 /// The item type returned when a [`WorldQuery`] is iterated over
 pub type QueryItem<'w, Q> = <Q as WorldQueryGats<'w>>::Item;
+/// The item type returned when a [`WorldQuery`] is iterated over during batching
```
```suggestion
/// The item type returned when a [`WorldQuery`] is iterated over in a batched fashion
```

InBetweenNames replied:
Done!

alice-i-cecile commented on crates/bevy_ecs/src/query/fetch.rs:
```diff
@@ -1097,13 +1260,45 @@ impl<T: Component> ChangeTrackers<T> {
 }
 
+#[derive(Clone)]
+pub struct ChangeTrackersBatch<'a, T, const N: usize, const ALIGN: usize>
```
Basic doc strings please for users stumbling across this.

alice-i-cecile commented on crates/bevy_ecs/src/query/fetch.rs:
```diff
@@ -1280,6 +1475,35 @@ unsafe impl<T: Component> WorldQuery for ChangeTrackers<T> {
+    ComponentTicks: AlignedBatchGat<N, ALIGN>,
+{
+    #[inline]
+    unsafe fn table_fetch_batched_aligned<'w>(
```
SAFETY comment needed.

InBetweenNames replied:
The safety constraints are documented in the `WorldQueryBatch`, similar to the safety constraints of `WorldQuery`'s `table_fetch` function -- is there something else that should be added here?

alice-i-cecile replied:
I'll defer to other reviewers on that point :)

alice-i-cecile commented on crates/bevy_ecs/src/query/fetch.rs:
```diff
@@ -1387,6 +1619,29 @@ macro_rules! impl_tuple_fetch {
+        impl<const N: usize, const ALIGN: usize, $($name: WorldQueryBatch<N, ALIGN>),*> WorldQueryBatch<N, ALIGN> for ($($name,)*)
+        {
+            #[inline]
+            unsafe fn table_fetch_batched_aligned<'w>(
```
SAFETY comment missing.

InBetweenNames replied:
Ditto here regarding `WorldQuery` and `table_fetch` (in case another reviewer comes across this)

alice-i-cecile commented on crates/bevy_ecs/src/query/mod.rs:
```diff
@@ -657,6 +661,115 @@ count():       {count}"#
+            let mut scalar_counter = 0;
+            let mut batch_counter = 0;
+
+            q.for_each_mut_batched::<4, 16, _, _>(
```
Needs a comment describing how these const generic values are set.

InBetweenNames replied:
Done!  More will be coming on this front too.  As it turns out, it was possible to use `impl` syntax to avoid having to pass in the `_` arguments for the lambdas.

alice-i-cecile commented on crates/bevy_ecs/src/query/state.rs:
```diff
@@ -759,6 +762,36 @@ impl<Q: WorldQuery, F: ReadOnlyWorldQuery> QueryState<Q, F> {
         }
     }
 
+    /// A read-only version of [`QueryState<Q,F>::for_each_mut_batched`].
```
Should explicitly call out that detailed docs can be found there.

alice-i-cecile commented on crates/bevy_ecs/src/query/state.rs:
```diff
@@ -778,6 +811,157 @@ impl<Q: WorldQuery, F: ReadOnlyWorldQuery> QueryState<Q, F> {
             );
         }
     }
+    /// This is a batched version of `for_each_mut` that accepts a batch size, `N`, and a desired alignment for the batches `ALIGN`.
```
```suggestion
    /// This is a batched version of [`for_each_mut`](Self::for_each_mut) that accepts a batch size, `N`, and a desired alignment for the batches `ALIGN`.
```

alice-i-cecile commented on crates/bevy_ecs/src/query/state.rs:
```diff
@@ -778,6 +811,157 @@ impl<Q: WorldQuery, F: ReadOnlyWorldQuery> QueryState<Q, F> {
         }
     }
+    /// This is a batched version of `for_each_mut` that accepts a batch size, `N`, and a desired alignment for the batches `ALIGN`.
+    /// Runs `f` and `f_batch` on each query result for the given [`World`].  This can be used to improve performance of your queries using SIMD.
```
These docs need to explain:
- what is alignment
- what is SIMD
- when this might improve (or worsen) performance

InBetweenNames replied:
Would it be okay if I just wrote a short blurb about these things, referring to e.g. Wikipedia or maybe some Intel blog posts for a more in-depth treatment?

alice-i-cecile replied:
Yep, that's totally fine. I just want this to be a bit less impenetrable to users who are new to low-level programming :)

InBetweenNames replied:
Hey @alice-i-cecile , I added a short little introduction to `Query::for_each_mut_batched` covering these concepts, would you mind giving it a read when you get a chance?  I don't want to go overboard with info but discoverability is a good thing.  Thank you!

alice-i-cecile commented on crates/bevy_ecs/src/query/state.rs:
```diff
@@ -778,6 +811,157 @@ impl<Q: WorldQuery, F: ReadOnlyWorldQuery> QueryState<Q, F> {
+    /// **Note**: Once generalized const expressions are stable, it will be possible to compute `ALIGN` automatically based on the component type and batch size.
+    /// Compile-time assertions will still exist for sanity checking.
+    ///  
+    /// In general, when using this function, be mindful of the types of filters being used with your query, as these can fragment your batches
```
Interesting advice, but needs to be a bit more concrete to be actionable.

InBetweenNames replied:
I'm thinking about putting an example here with change trackers -- the query should still be "dense" but possibly highly fragmented.  The effect of that is that the "scalar path" has to run a lot more often due to the incomplete batches available.

For example (forgive the ASCII art):

~~~
batch =   | b b b b b b b b |
filter => | y y y y n y y y | 

~~~

In this example, a would-be batch is "fragmented" because the filter yielded `false` for one element of the batch.  Therefore, the scalar path has to be executed for all elements of the batch, which is no better than `for_each_mut`.

But a more friendly query filter might be something like `With<Component>`, which basically just restricts which tables are included in the query.  That is, the query itself isn't "fragmented" anymore.  Elements won't just spuriously return `false` in the middle of iteration of a table -- the filter will either succeed for the entire table, or fail for the entire table.

I believe these are considered "Archetypal" queries in Bevy but I haven't found much documentation on that.

Btw, even with "friendlier" filters, the initial batching implementation isn't that smart yet... I expect performance to not be super great with them at the moment.  However, the interface is at least built such that code using that would get faster as better implementations are made.  For now I think the "avoid filters" advice is the most practical, until we get those smarts into the batch iteration code anyway.

alice-i-cecile replied:
Good, I think this is a great start.

alice-i-cecile commented on crates/bevy_ecs/src/query/state.rs:
```diff
@@ -778,6 +811,157 @@ impl<Q: WorldQuery, F: ReadOnlyWorldQuery> QueryState<Q, F> {
+    /// #[repr(transparent)]
+    /// struct Velocity(Vec3);
+    ///
+    /// //Important: this allows us to "cast away" the outer layer of Velocity
```
Safety comment needed.

alice-i-cecile replied:
(and above)

alice-i-cecile commented on crates/bevy_ecs/src/query/state.rs:
```diff
@@ -778,6 +811,157 @@ impl<Q: WorldQuery, F: ReadOnlyWorldQuery> QueryState<Q, F> {
+    /// unsafe impl TransparentWrapper<Vec3> for Position {}
+    ///
+    /// #[derive(Component, PartialEq, Debug)]
+    /// #[repr(transparent)]
```
Explain why we want `repr(transparent)`

alice-i-cecile commented on crates/bevy_ecs/src/storage/aligned_vec.rs:
```diff
@@ -0,0 +1,258 @@
+/// Intended to support SIMD use cases.
+///
+/// Used to densely store homogeneous ECS data whose type is known at compile time.
+/// Built on `BlobVec`.  It is not intended to be a drop-in replacement for Vec at this time.
```
```suggestion
/// Built on `BlobVec`. It is not intended to be a drop-in replacement for Vec at this time.
```

InBetweenNames commented:
Wow, you're fast!!  I will see about getting these finished in the coming days, thanks for your patience :)

InBetweenNames commented on crates/bevy_ecs/src/change_detection.rs:
```diff
@@ -271,6 +283,143 @@ change_detection_impl!(Mut<'a, T>, T,);
 impl_debug!(Mut<'a, T>,);
 
+/// Unique mutable borrow of an entity's component (batched version)
+/// a batch changes in unison.  A batch has changed if any of its elements have changed.
```
Done!

InBetweenNames commented:
Happy Thanksgiving!

Hah, I'm glad I wrote tests.  Found a few soundness issues thanks to the Miri CI we have here.  You bet I'll be running that locally now too :) fortunately the issues weren't fundamental.

Currently working on the remaining documentation related problems now that the soundness issues have been resolved.

There's a few ways of using the API and the current example only shows the more "advanced" way with repr(transparent).  It's now possible to more easily use batches with a non-"transparent" API using the `map` and `as_array` functions of the `AlignedBatch` trait.  The tradeoff is that it seems to optimize worse -- the [Rust docs](https://doc.rust-lang.org/std/primitive.array.html#method.map) seem to indicate that the performance situation is a little finicky with array maps.  In a quick test I did, I saw loads and stores that could have been easily optimized out appearing in the assembly.  The docs recommend using the `iter()` API instead -- maybe I'll play around with that later this week after I clean up the docs a bit more.

BoxyUwU commented on crates/bevy_ecs/src/query/fetch.rs:
```diff
@@ -444,6 +448,49 @@ pub trait WorldQueryGats<'world> {
+    ) -> <Self as WorldQueryBatchGats<'w, N, ALIGN>>::FullBatch;
+}
+
+impl<const N: usize, const ALIGN: usize> WorldQueryBatch<N, ALIGN> for Entity
```
this impl should be with the other worldquery impls for `Entity`, not in the middle of a bunch of trait and type alias definitions

InBetweenNames replied:
Yes, totally agree, I will see about organizing this better and possibly moving it to a new file for clarity.  Thank you!

InBetweenNames replied:
@BoxyUwU I've improved this a bit by reorganizing `fetch.rs` to always have the definitions in the same order for each type of query.  However, I might move these definitions to a new file with the generalized const expressions changes coming up.

InBetweenNames replied:
Marking this resolved as the GAT refactor cleaned all of this right up.  Still might move things to a new file for clarity.

BoxyUwU reviewed (changes requested):
I havent reviewed all of this, I don't think we should merge this with the current changes to `fetch.rs` but I also think there is not much we can do until `feature(generic_associated_types)` is stabilized (other than removing the genericness over `ALIGN` and having 3 separate fns or just having one). next rust release should have stable gats which is in ~20 days so I don't think this is too much of an ask to wait (and its extraordinarily unlikely imo that this PR actually gets merged in such a short timeframe anyway)

InBetweenNames commented:
Thanks very much for your comments @BoxyUwU -- this is a learning experience in Rust for me and I'll work at addressing yours (and the rest of @alice-i-cecile 's ) comments over the next few weeks.  I think delaying until GATs are stable makes sense and I can actually start readying the code for that in the meantime.  Thank you for your patience :)

InBetweenNames commented on crates/bevy_ecs/src/storage/blob_vec.rs:
```diff
@@ -49,10 +49,37 @@ impl BlobVec {
+                Therefore we have the following check:
+        */
+
+        if item_layout.align() > batch::MAX_SIMD_ALIGNMENT {
+            panic!(
+                "Attempted to create a BlobVec with an item layout alignment > MAX_SIMD_ALIGNMENT"
+            );
+        }
```
Thanks for the feedback -- I had code like this originally and Miri was throwing a warning, but since you're cool with `-Zmiri-permissive-provenance` I'll go ahead and make that change.

BoxyUwU replied:
CI already runs with `-Zmiri-permissive-provenance` ideally we'd use `ptr::invalid(item_layout.align().max(64))` but thats unstable so oh well :sweat_smile:

InBetweenNames replied:
Done, thank you!

InBetweenNames commented on crates/bevy_ecs/src/storage/blob_vec.rs:
```diff
@@ -353,6 +380,9 @@ impl Drop for BlobVec {
 fn array_layout(layout: &Layout, n: usize) -> Option<Layout> {
     let (array_layout, offset) = repeat_layout(layout, n)?;
     debug_assert_eq!(layout.size(), offset);
+
+    //Note: NEEDED for batching.  This is the layout of the array itself, not the layout of its elements.
+    let array_layout = array_layout.align_to(batch::MAX_SIMD_ALIGNMENT).unwrap();
```
Sorry I can see where I wasn't clear with this.  `stride == size` definitely still holds -- this statement just aligns the *start* of the array, if that makes sense (the chunk of memory for the array *itself* is aligned to `MAX_SIMD_ALIGNMENT`).  The element layout remains exactly the same as it was before.  This was needed to guarantee that a batch starting at index 0 would be aligned.

For interested readers, typically SIMD code operates in three "phases", a prologue, a "vector region", and an epilogue.
The prologue and epilogue are just "scalar code paths", i.e., your typical for loops.  The prologue's purpose is to process elements up until it is possible to execute vector instructions, which generally require a certain alignment to work.  However, the vector region can only execute in batches of size `N`, which means that if there are any remaining items to do that can't fit in a batch of size `N`, they have to be processed serially.


As an example, consider an array of `[f32; 128]`, which is aligned to a boundary of 4 (say, address `100`).  Suppose you had a special vector instruction set that could operate on batches of 8 `f32`s at a time, the catch being, the loads have to be aligned to 32 as well (sizeof `f32` being `4`).  Then, because of the alignment of the array, you would only be able to batch from index 8 onwards.  Then you would of course need a prologue to handle that "slow path".  However, if the array started at address `96`, you could see that no prologue would be necessary.

This line here eliminates the prologue case from Bevy by guaranteeing that batch 0 is aligned to the correct SIMD alignment in all cases, because `MAX_SIMD_ALIGNMENT` is the greatest alignment required of any instruction on the given system (including vector instructions).  Does this make sense?  Thank you!

BoxyUwU replied:
stride == size does not hold for the layout returned by this fn, `align(MAX_SIMD_ALIGNMENT) size(2)` for example isnt stride == size but could be potentially returned from this fn

InBetweenNames replied:
Yes that's exactly what I was after -- this `array_layout` is what is ultimately used to allocate memory for the entire array *after* the stride has already been taken into account.  It is derived from the `item_layout` using the `repeat_layout` function.  The flow (as I understand it) works as follows:

1 - The `item_layout` of `T` is passed in via the `BlobVec` constructor and held. (stride matters here, as you rightly point out!)
2 -  The `array_layout` is generated from `repeat_layout(layout, n)` (stride is taken into account by repeat_layout -- this gives us a Layout that can hold N `item_layout`s [size == stride] and aligned to alignment of item_layout)
3 -  The "SIMD aligned" `array_layout` is obtained by taking that `array_layout` and then calling `align_to` on it with `MAX_SIMD_ALIGNMENT` (stride has already been taken into account -- we are simply telling the allocator "take this Layout and ensure that the alignment requirement is `MAX_SIMD_ALIGNMENT`.  I.e. this is now a Layout that can hold N `item_layout`s [size == stride] except the the chunk of memory will now be aligned to MAX_SIMD_ALIGNMENT)
4 - It is ultimately (3) that is passed into the allocator

Note that the `item_layout` is left pristine through all of this to ensure we don't add any extra padding between elements.  It is strictly just the `array_layout` that gets `align_to(MAX_SIMD_ALIGNMENT)`.
The actual elements of the array are still laid out like `repeat_layout` would do it.

This is the best of my understanding of that at the moment.  If the code called `align_to` on the `item_layout` itself, then `stride != size` for sure.  But since it's being called on `array_layout` *after* `repeat_layout` has already done its job, the stride should already have been taken into account.  Is my understanding right here?

(Apologies for the ASCII art, I'm a visual thinker :D)

~~~

1 - item_layout provided to constructor
2 - align(item_layout):        | item_layout item_layout item_layout item_layout ... (N times) | (repeat_layout)
3 - align(MAX_SIMD_ALIGNMENT): | item_layout item_layout item_layout item_layout ... (N times) | (array_layout)

~~~

InBetweenNames replied:
A rust playground link illustrating it a little further: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=099b1168135ed7ed189dd3bad76477c1

InBetweenNames replied:
Since it's been a while, I'm just posting an [updated gist link with code embedded](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&code=use%20core%3A%3Aalloc%3A%3ALayout%3B%0A%0Aconst%20MAX_SIMD_ALIGNMENT%3A%20usize%20%3D%2064%3B%0A%0A%2F%2F%2F%20From%20%3Chttps%3A%2F%2Fdoc.rust-lang.org%2Fbeta%2Fsrc%2Fcore%2Falloc%2Flayout.rs.html%3E%0Afn%20simd_array_layout(layout%3A%20%26Layout%2C%20n%3A%20usize)%20-%3E%20Option%3CLayout%3E%20%7B%0A%20%20%20%20let%20(array_layout%2C%20offset)%20%3D%20repeat_layout(layout%2C%20n)%3F%3B%0A%20%20%20%20debug_assert_eq!(layout.size()%2C%20offset)%3B%0A%0A%20%20%20%20%2F%2FNote%3A%20NEEDED%20for%20batching.%20%20This%20is%20the%20layout%20of%20the%20array%20itself%2C%20not%20the%20layout%20of%20its%20elements.%0A%20%20%20%20let%20array_layout%20%3D%20array_layout.align_to(MAX_SIMD_ALIGNMENT).unwrap()%3B%0A%20%20%20%20Some(array_layout)%0A%7D%0A%0A%2F%2F%2F%20From%20%3Chttps%3A%2F%2Fdoc.rust-lang.org%2Fbeta%2Fsrc%2Fcore%2Falloc%2Flayout.rs.html%3E%0A%2F%2F%20Original%20unmodified%20array_layout()%20function%0Afn%20original_array_layout(layout%3A%20%26Layout%2C%20n%3A%20usize)%20-%3E%20Option%3CLayout%3E%20%7B%0A%20%20%20%20let%20(array_layout%2C%20offset)%20%3D%20repeat_layout(layout%2C%20n)%3F%3B%0A%20%20%20%20debug_assert_eq!(layout.size()%2C%20offset)%3B%0A%0A%20%20%20%20Some(array_layout)%0A%7D%0A%0A%2F%2F%20TODO%3A%20replace%20with%20%60Layout%3A%3Arepeat%60%20if%2Fwhen%20it%20stabilizes%0A%2F%2F%2F%20From%20%3Chttps%3A%2F%2Fdoc.rust-lang.org%2Fbeta%2Fsrc%2Fcore%2Falloc%2Flayout.rs.html%3E%0Afn%20repeat_layout(layout%3A%20%26Layout%2C%20n%3A%20usize)%20-%3E%20Option%3C(Layout%2C%20usize)%3E%20%7B%0A%20%20%20%20%2F%2F%20This%20cannot%20overflow.%20Quoting%20from%20the%20invariant%20of%20Layout%3A%0A%20%20%20%20%2F%2F%20%3E%20%60size%60%2C%20when%20rounded%20up%20to%20the%20nearest%20multiple%20of%20%60align%60%2C%0A%20%20%20%20%2F%2F%20%3E%20must%20not%20overflow%20(i.e.%2C%20the%20rounded%20value%20must%20be%20less%20than%0A%20%20%20%20%2F%2F%20%3E%20%60usize%3A%3AMAX%60)%0A%20%20%20%20let%20padded_size%20%3D%20layout.size()%20%2B%20padding_needed_for(layout%2C%20layout.align())%3B%0A%20%20%20%20let%20alloc_size%20%3D%20padded_size.checked_mul(n)%3F%3B%0A%0A%20%20%20%20%2F%2F%20SAFETY%3A%20self.align%20is%20already%20known%20to%20be%20valid%20and%20alloc_size%20has%20been%0A%20%20%20%20%2F%2F%20padded%20already.%0A%20%20%20%20unsafe%20%7B%0A%20%20%20%20%20%20%20%20Some((%0A%20%20%20%20%20%20%20%20%20%20%20%20Layout%3A%3Afrom_size_align_unchecked(alloc_size%2C%20layout.align())%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20padded_size%2C%0A%20%20%20%20%20%20%20%20))%0A%20%20%20%20%7D%0A%7D%0A%0A%2F%2F%2F%20From%20%3Chttps%3A%2F%2Fdoc.rust-lang.org%2Fbeta%2Fsrc%2Fcore%2Falloc%2Flayout.rs.html%3E%0Aconst%20fn%20padding_needed_for(layout%3A%20%26Layout%2C%20align%3A%20usize)%20-%3E%20usize%20%7B%0A%20%20%20%20let%20len%20%3D%20layout.size()%3B%0A%0A%20%20%20%20%2F%2F%20Rounded%20up%20value%20is%3A%0A%20%20%20%20%2F%2F%20%20%20len_rounded_up%20%3D%20(len%20%2B%20align%20-%201)%20%26%20!(align%20-%201)%3B%0A%20%20%20%20%2F%2F%20and%20then%20we%20return%20the%20padding%20difference%3A%20%60len_rounded_up%20-%20len%60.%0A%20%20%20%20%2F%2F%0A%20%20%20%20%2F%2F%20We%20use%20modular%20arithmetic%20throughout%3A%0A%20%20%20%20%2F%2F%0A%20%20%20%20%2F%2F%201.%20align%20is%20guaranteed%20to%20be%20%3E%200%2C%20so%20align%20-%201%20is%20always%0A%20%20%20%20%2F%2F%20%20%20%20valid.%0A%20%20%20%20%2F%2F%0A%20%20%20%20%2F%2F%202.%20%60len%20%2B%20align%20-%201%60%20can%20overflow%20by%20at%20most%20%60align%20-%201%60%2C%0A%20%20%20%20%2F%2F%20%20%20%20so%20the%20%26-mask%20with%20%60!(align%20-%201)%60%20will%20ensure%20that%20in%20the%0A%20%20%20%20%2F%2F%20%20%20%20case%20of%20overflow%2C%20%60len_rounded_up%60%20will%20itself%20be%200.%0A%20%20%20%20%2F%2F%20%20%20%20Thus%20the%20returned%20padding%2C%20when%20added%20to%20%60len%60%2C%20yields%200%2C%0A%20%20%20%20%2F%2F%20%20%20%20which%20trivially%20satisfies%20the%20alignment%20%60align%60.%0A%20%20%20%20%2F%2F%0A%20%20%20%20%2F%2F%20(Of%20course%2C%20attempts%20to%20allocate%20blocks%20of%20memory%20whose%0A%20%20%20%20%2F%2F%20size%20and%20padding%20overflow%20in%20the%20above%20manner%20should%20cause%0A%20%20%20%20%2F%2F%20the%20allocator%20to%20yield%20an%20error%20anyway.)%0A%0A%20%20%20%20let%20len_rounded_up%20%3D%20len.wrapping_add(align).wrapping_sub(1)%20%26%20!align.wrapping_sub(1)%3B%0A%20%20%20%20len_rounded_up.wrapping_sub(len)%0A%7D%0A%0A%2F%2FS%20should%20have%20align_of(i32)%0Astruct%20S%0A%7B%0A%20%20%20%20x%3A%20i32%0A%7D%0A%0Afn%20main()%0A%7B%0A%20%20%20%20let%20s_layout%20%3D%20Layout%3A%3Anew%3A%3A%3CS%3E()%3B%0A%0A%20%20%20%20let%20simd_blob_vec_layout%20%3D%20simd_array_layout(%26s_layout%2C%2053).unwrap()%3B%0A%20%20%20%20%0A%20%20%20%20let%20blob_vec_layout%20%3D%20original_array_layout(%26s_layout%2C%2053).unwrap()%3B%0A%20%20%20%20%0A%20%20%20%20println!(%22simd_blob_vec_layout%3A%20%7B%3A%3F%7D%22%2C%20simd_blob_vec_layout)%3B%0A%20%20%20%20println!(%22blob_vec_layout%3A%20%7B%3A%3F%7D%22%2C%20blob_vec_layout)%3B%0A%20%20%20%20println!(%22simd_blob_vec_layout.size()%20%3D%3D%20simd_blob_vec_layout.size()%3A%20%7B%7D%22%2C%20simd_blob_vec_layout.size()%20%3D%3D%20blob_vec_layout.size())%3B%0A%20%20%20%20%0A%20%20%20%20%2F%2FNote%20that%20the%20sizes%20of%20the%20above%20layouts%20are%20identical!%0A%20%20%20%20%0A%20%20%20%20%0A%7D) for when we come back around to discussing this. in case the original link fails to work.

InBetweenNames commented on crates/bevy_ptr/src/batch.rs:
```diff
@@ -0,0 +1,400 @@
+            //A "batch" of ZSTs is itself a ZST.
+            //This sounds dangerous, but won't cause harm as nothing
+            //will actually access anything "in the slice".
+            //ZSTs are trivially well aligned, so there is nothing to be checked.
```
Definitely an oversight on my part, and I agree this needs to be done better.  Thank you!

InBetweenNames replied:
Yeah the comment about ZSTs being trivially well aligned is wrong.  I've removed that.  Fortunately it didn't have much to do with the rest of the code :)

Do you have any suggestions for how to check that the pointer is aligned idiomatically?  [This StackOverflow](https://stackoverflow.com/a/71982312) seems to suggest that `align_offset` is the way to go until `pointer_is_aligned` is stable.  I've rewritten this as a pointer to integer cast modulo `ALIGN` (with a TODO to change this when `pointer_is_aligned` is stable), but I'm not 100% convinced this is the best way to go either.

The latest commit I've pushed up has these changes.

InBetweenNames commented:
Next up on my list is to get one of the Bevy examples going with batched queries (in a very limited way) to demonstrate batched queries a bit better.  Thanks for your comments and your patience everyone.

alice-i-cecile commented:
Not a problem! It's been great to see this slowly progress.

james7132 commented:
@InBetweenNames can you please mark in the PR description that this `Fixes #6161` so that it gets auto-closed when this is merged?

InBetweenNames commented:
> @InBetweenNames can you please mark in the PR description that this `Fixes #6161` so that it gets auto-closed when this is merged? I'll give this a review soon. This has been quite the API hole compared to what Unity's ECS has that Bevy sorely needs.

Not a problem!  I have some more changes coming down the pipeline when I get the energy to make the API a little friendlier.  With the stabilization of GATs being imminent, I'd like to clean up the code with those (not a big change), and also restructure `for_each_mut_batched` to work similar to the following:

```rust
    for_each_mut_batched::<4>( |((mut a, Align16), (b, Align32))| {
        
        println!("a: {:?} b: {:?}", a, b);
        
    } );
```

That is, the `Align` parameter now becomes per-component, which is way more flexible.  As a bonus, I can provide a `for_each_mut_batched` variant that doesn't care about alignment -- with the usual caveats that would imply in terms of codegen.

For the above example, such a variant would be equivalent to:

```rust
    for_each_mut_batched::<4>( |((mut a, Unaligned), (b, Unaligned))| {
        
        println!("a: {:?} b: {:?}", a, b);
        
    } );
```

Simple, safe, effective :)

Can't promise any timelines for that (hopefully this month), but I am working on it in the evenings here.  Bevy is in a very unique position to provide HPC primitives like this :)  Down the pipeline, we can look at avoiding false-sharing in a type-safe and compile time checked manner when doing batching across multiple threads, too.  That's my dream toolkit for simulations right there :)

InBetweenNames commented:
Alright, managed to get this ported over to the new GAT way of doing things and simplified a lot in the process.  However, the per-component alignment feature isn't implemented yet.  I'm going to try seeing if I can use `typenum` to automatically compute it, which would keep things still relatively simple while still not relying on generic const expressions (which seem to have some foundational problems).  I've moved the PR over to "draft" status as I don't think it should be merged before alignments can be manually set per component or they can be automatically computed.  Still, it's worth reviewing the changes so far as what is committed shouldn't change drastically after those "feature complete" changes are in.

InBetweenNames commented on crates/bevy_ptr/src/lib.rs:
```diff
@@ -195,29 +204,60 @@ impl<'a> OwningPtr<'a> {
+            //This sounds dangerous, but won't cause harm as nothing
+            //will actually access anything "in the slice"
+            debug_assert!(ptr.align_offset(batch::MAX_SIMD_ALIGNMENT) == 0);
+        }
```
(See comment above)

InBetweenNames commented:
I took a little detour to try out what this would look like if `generic_const_exprs` were stable, and it's actually really nice:

```rust

q.for_each_mut_batched::<4>(
                |(mut position, velocity)| { 
                    //Scalar path
                },
                |(mut position, velocity)| { 
                    //Batched path
                });
```

Code provided here: https://github.com/inbetweennames/bevy/tree/generic_const_exprs

All alignments automatically computed, absolutely zero work needs to be done by the user to do anything, and the autovectorizer loves it.  The type is very simple to understand too: `AlignedBatch<T, N>` automatically has the correct batch alignment computed using the `gcd`!  Literally all the user needs to do is select the batch size `N`, provide a scalar and vector path, and the rest is taken care of for you.  It really is that nice.

(One can dream!)

Obviously I can't propose that for merging, but it does serve as an important design validation point that you really can make some trivial changes to the code once it is possible to evaluate const expressions in generics and just have this for free.  I suppose this could be hacked in with `typenum` too, maybe... might be worth a shot.

InBetweenNames commented:
Bad news :( ... even with the `typenum` hackery, it seems this can't be done on stable as it's not possible to get the size of a type `T` as a typenum without using `generic_const_exprs`.  Therefore, it seems the best we can do (for now, on stable) is implement the per-component manual alignments as described [previously](https://github.com/bevyengine/bevy/pull/6161#issuecomment-1299323904).  That's okay, because the "unaligned" variant can then be promoted to "alignments automatically derived" without code changes once `generic_const_exprs` are stable, and the manual aligned version can easily migrated off.  I'll go ahead and start working on that.  The API is still very usable this way :)

InBetweenNames commented:
I was wrong!

It is indeed possible to do this with `typenum`, but it makes the code a giant mess, and likes to cause internal compiler errors in `rustc`.  It requires a patched `typenum` that supports const generics (a `U<N> = UN` type) and a new trait, `SizeOfAsTypenum`, which can be conveniently hacked into the derive macro for `Component`.  `SizeOfAsTypenum` defines a type `Output` that is set by the implementor to be the size of the type as a typenum.  Since this is not a generic context, it is indeed possible to use sizeof here :) (it looks like `type Output = U< { sizeof::<T>() }>`, hence the dependency on the patched `typenum`).  `SizeOfAsTypenum` needs to be manually implemented for other, non-component types, which isn't great (and attempting to provide an implementation generically will, of course, require `generic_const_exprs`).

However, even without the ICEs, this is still a gigantic hack and (I expect) unlikely to be accepted.  I think given the options of pushing a little bit of complexity onto the user versus putting a lot of complexity in Bevy to automatically derive the minimum alignments (on stable Rust), it makes sense to prefer the former path.  This still provides for a clean upgrade path when `generic_const_exprs` are stable that reduces both user and implementation complexity compared to the manual alignment approach.

alice-i-cecile commented:
I agree with that assessment!

InBetweenNames commented:
> iamthetypesystem.jpg

Small update -- I was unable to get to the proposed implementation without (again) hitting the limits of the type system in Rust as it currently exists.  So the proposed:

```rust

    for_each_mut_batched::<4>( |((mut a, Align16), (b, Align32))| {
        
        println!("a: {:?} b: {:?}", a, b);
        
    } );
```

isn't implementable without `generic_const_exprs`, which if those were stable, we'd just use the automatic approach that already has been proven to work in the branch I discussed earlier.  I'll discuss the exact limitation of Rust's type system in a moment that gets us in trouble, but first I want to discuss what I think is a suitable compromise:

```rust

            //Align32, Align16, etc, are just type synonyms for Align<N> from the `elain` crate
            //type Unaligned = Align<1>;
            q.for_each_mut_batched::<4, (Align32, Align8, Align16, Unaligned)>(
                |(mut a, b, c, d)| {
                    //scalar
                },
                |(mut a, b, c, d)| {
                    //batched
                });
```

The above works just fine and is what I'm going to propose for merging after I clean it up a bit and move the definitions to a new file.  The main difference is that the alignments can't be passed inline with the arguments as before -- they must be passed as a generic parameter.  Unfortunately, Rust's type system again strikes here, where it is not possible to default this to "`Unaligned`" (i.e. `Align<1>`).  The reason for this is discussed in this thread (most recent comment linked):

https://github.com/rust-lang/rust/issues/36887#issuecomment-1041515594

However, another approach is easily worked around for this: we provide a new function where this is defaulted using another associated type, for the unaligned case.  That machinery is pretty unobtrusive, even if it is a bit of a hack.  Unfortunately, that machinery is more verbose than it needs to be due to `associated_type_defaults` again not being stable lol.  Still, it's not too bad.
If you want the power of stipulating manual alignments because you need that, it is in your hands -- otherwise, no need to worry about it.  This should help discoverability of batching and vectorization in general.

On the topic of discoverability, what I propose is to make the "Unaligned" interface the default for simplicity, as there's only a handful of architectures that really care about alignment hints, or have instruction opcodes that demand aligned operands.
For example, `x86-64-generic" targets SSE2, and a good chunk of SSE2 instructions demand naturally aligned operands.  Therefore, without being able to prove natural alignment, the compiler must emit more `movup*` instructions for the generic x86-64 target and do work in registers, which will hurt performance.  So, `Align<16>` clearly helps here.  Even though newer processors don't make a distinction between "aligned" and "unaligned" variants of instructions (the "unaligned" variant performs as fast as "aligned" so long as the data is actually well aligned), you still pay a price.  Older processors will of course pay a much steeper price.   As another example, `armv7a` with NEON also accepts alignment hints (which are free to be ignored by the implementation).

#### Truly modern architectures (things made in the last 10 years)

Of course, truly modern x86 (AVX support and above) makes the full instruction set available for unaligned operands thanks to VEX/EVEX encodings.  AArch64 in a similar vein has no "aligned" variants of instructions at all.  In both cases, full performance is achieved if the memory operands are naturally aligned, but the instructions don't need to encode that alignment themselves -- meaning there is nothing for your compiler to really do here.  We could simply ignore alignment and follow suit, simplifying this PR, but I want to stress that AVX is not in the default enabled instruction sets for  `x86-64`.  Even if we move the `x86-64-v2`, that only enables SSE4_2.  It takes `x86_64-v3` and above to enable AVX support in the compiler, and so far, I've only seen discussion about moving the default x86-64 target to `v2`.  Also, if Bevy were to ever run on embedded platforms, you'd really want this to eek out every last drop of performance possible.  Or, maybe running Bevy systems someday as GPU kernels?



#### Type madness

I saved this part for the end to not have to inflict it on poor souls casually passing by.  You've been warned!

The exact limitation I ran into when trying to implement the former interface is that Rust GATs unfortunately only mirror Haskell's "type families" (which represent "injective" relationships) and there is no way to assert that the combination of alignment specifiers is "unique" with respect to a trait.  In formal language terms, you say there is a possibility of "ambiguity" here.  The other side of higher kinded types is the ability to model bijective relationships, where you can indeed make that assertion and "work backwards".  At the call site, when we say, for example, |Batch16(mut a), Batch8(b), Batch4(c), Batch32(d)|, without that kind of functional dependency (sometimes called fundep in Haskell space), there is no way to assert to Rust that there is exactly one appropriate derivation for that sequence of arguments.  Things get even more complicated when you consider weird queries like `Query<(&A, (&B, (&C, &D)))>`, which are nested.  Even if you enforce a flat hierarchy by introducing a new trait (say, `WorldQueryBatchBase`) to exclude this possibility, having `WorldQueryBatch` impls for the query tuples depend on those , without `generic_const_exprs` you can't have `WorldQueryBatchBase::fetch_batched` return some kind of `BatchT<Self::MIN_ALIGN, Self::FullBatchType>`, as `MIN_ALIGN` would have to be chosen by the implementor 🙃...

#### Conclusion

So... I'm fairly confident that on stable rust, the new proposed interface at the start of this post is the best we can do, with an "unaligned" and a separate "manually aligned" interface for the use cases outlined above.  When `generic_const_exprs` are stable someday, we can eliminate the "manually aligned" variant and automatically upgrade users of the "unaligned" variant to the "automatically aligned" version without breaking any backwards compatibility as the interface is the exact same.

Until next time!

InBetweenNames commented:
Small update; code has been reworked, but I'm trying to find a way to recommend using batches without using the array `map` function, which optimizes horribly for some reason.  TransparentWrapper from `bytemuck` may be the way to go, but I'll be exploring options over the coming days.  If anyone has ideas on how to do this, your advice would be most welcome.   It may be that we just recommend people to unroll their batches manually such as `let [Position(p1), Position(p2), ..., Position(pn)] = batch` to handle these kinds of cases.

InBetweenNames commented:
So it seems `array::map` can be fixed, and there is work upstream on doing that: [here](https://github.com/rust-lang/rust/issues/86912) and [here](https://github.com/rust-lang/rust/issues/102202).  It is also possible to provide a new trait to use the proposed implementations that fix these problems, e.g. `ArrayMap`.  I found in my own tests, taking the array by reference instead of by value (move) was sufficient for the compiler to generate sane code for a `map` function.  Of course, this is solely to work around the poor performance of `array::map`.  Since it seems like the upstream implementation is likely to be fixed in the next 6 months, I'm going to go forward and recommend users unroll their batches manually as mentioned before.  There's no sense in introducing a new trait due to an upstream bug IMO, which would be obsolete as soon as it's fixed.

It's curious that the upstream API takes the array by move -- it seems there's no real benefit to doing that and it makes the function harder to optimize.  I should try to dig up some of the discussion around `array_map` and figure out what's going on there.

With that said, all that's left to do is update the examples and docs and satisfy the CI here.  At which point, I'll mark this again for review.  Shouldn't be too long now :)

InBetweenNames commented:
I have an idea for a new example to showcase this PR, but I haven't had time to get it implemented with benchmarks and everything.   However, I think the actual code itself is ready for review.

A universal "batching is always better" example is very hard to write, because hardware isn't universal.  The best I've been able to come up with, in stable rust at least, is using x86 AVX intrinsics directly to perform a layout transform and doing some math that way.  AVX intrinsics are not beginner friendly though, and I'm concerned about putting those into an example that is intended to be beginner friendly.  `std::simd` helps here, but it isn't available in stable Rust unfortunately -- and has it's own problems (that appear to be being worked on).

I think some properties for a good enough example would be:

* Has to be realistic, not contrived in any way: e.g., collision detection, some kind of physics, etc
* If it uses `core::arch` intrinsics, has to be relatively portable, and can't use anything "too deep"
  * If it doesn't, and it relies on autovectorization, then it must be very obvious, because rustc/LLVM is fickle between versions on what it will autovectorize and what it won't.
* It can't be too complicated: ideally it shouldn't define a bunch of types, etc, and should be short
* It shouldn't be worse than letting `rustc` autovectorize everything.

So far, I have a contrived example involving explosions and damage calculations, but I fear it's not super realistic with what you might actually see in a game.  I was thinking maybe a collision detection off a solid plane would make more sense to implement, and might not be too much longer.

For now, I took out the old example because it wasn't very good -- on recent `rustc`s, it got unrolled and autovectorized optimally, making batching actually perform worse as `rustc` couldn't optimize the batched version as easily.

From this point onwards I'll focus on the documentation and seeing what we can do for a cool "intro example" for vectorized queries.

InBetweenNames commented:
Hey all, just lettying you know I haven't forgotten about this.  I've found a suitable example using collision detection that is easy to explain and directly benefits from batched queries.  It's simply doing an inside/outside test of a convex polygon with respect to a set of entities.  So we've got some dot products, comparisons, etc, but nothing crazy, and it's something you might expect to see in an actual game.  Currently working on getting that written up in a non-ugly way for inclusion.  Unfortunately, due to `std::simd` being unstable, I have to write it using `core::arch` intrinsics.  I hope that's not a dealbreaker!

On a side note, Rust seriously needs some kind of copy elision baked into the standard:

![image](https://user-images.githubusercontent.com/7820684/211227882-c48ab3d6-8707-48c7-9105-ae33adfde18b.png)

What you're seeing here is literally the following:

> "copy this crap to my stack, then load the stuff i just copied into my stack into registers"

This seems to be a similar problem as we've seen with `array::map`, where copies aren't elided when they could be.  In C++, we have guaranteed copy elision which helps in some cases as this gets enforced in the language frontend.  In Rust, it seems that the lower levels of the stack are relied upon to get rid of extraneous copies, but it's a little flakey.

Otherwise, even with that shortcoming, the code still outperforms its scalar and autovectorized counterparts.

On the topic of benchmarks, it's my hope that I can move the ones out of my `bevy_simd_bench` repo into the main tree without much fuss.

alice-i-cecile commented:
> On the topic of benchmarks, it's my hope that I can move the ones out of my bevy_simd_bench repo into the main tree without much fuss.

Excellent: these will be much easier to merge in their own PR.

InBetweenNames commented:
The new example and associated benchmark are up.  Results look good!

![image](https://user-images.githubusercontent.com/7820684/212591512-be6ff800-e0f3-4c82-af92-67cbf083c258.png)

This benchmark has been included in the main Bevy benchmark suite.  To run it, ensure you have AVX enabled -- the suite will exclude it it isn't enabled.

InBetweenNames commented:
@alice-i-cecile @james7132  @BoxyUwU @TheRawMeatball I think this is now fully ready for review!

For example usage, see [this benchmark](https://github.com/bevyengine/bevy/pull/6161/files#diff-dcc0f753bd330ada950bb7791f6dc51175417abec661b11013f77a461c6aa816) 😄

alice-i-cecile commented:
Awesome :D I'll aim to do a full review by the end of next Monday: there's a *ton* of ECS work in the air right now.

InBetweenNames commented:
Sounds good, no rush! :)

InBetweenNames commented:
No changes in the latest commit; just rebased.  I see some new verbiage in main about "BatchedStrategies" for parallel queries.  That's cool -- just wondering if maybe a different name should be given to this different concept introduced in this PR to avoid confusion.  While similar, the intent is different: the batching introduced in this PR is about processing batches of N components at a time using SIMD, whereas parallel batch sizes appear to be more about managing overhead and synchronization costs when executing a query across multiple threads.  I'm open to suggestions here :)

InBetweenNames commented:
When #6773 is merged, I'll port this over to using the new style of iteration.

InBetweenNames commented:
FYI: did a little reading into the naming convention behind this concept.  EnTT refers to it as "chunked iteration": https://github.com/skypjack/entt/issues/462

How do we feel about using the name "chunk" instead of "batch" for this PR, to set it apart from other uses of "batch" in Bevy?  Is there a better name we could use instead?

alice-i-cecile commented:
> [7:27 AM] james7132: On closer inspection, I don't think the batched queries is going to be ready for 0.11 (or even 0.13 at this rate), it's reliance on unstable rust features is probably a non-starter for it.

From @james7132 on [Discord](https://discord.com/channels/691052431525675048/749335865876021248/1080103960943267911). If you can find a way to make it work without this, or convince us that the assessment is wrong / it's worth making an exception, let us know.

Similarly, if you need us to make changes to the internals to allow 3rd party crates that need unstable Rust features to implement this externally, we're also generally very open to that!

InBetweenNames commented:
Hey Alice, this was an old comment that was addressed already. This code
does not rely on any unstable rust features. Thank you!

On Tue, Feb 28, 2023, 8:22 a.m. Alice Cecile ***@***.***>
wrote:

> [7:27 AM] james7132: On closer inspection, I don't think the batched
> queries is going to be ready for 0.11 (or even 0.13 at this rate), it's
> reliance on unstable rust features is probably a non-starter for it.
>
> From @james7132 <https://github.com/james7132> on Discord
> <https://discord.com/channels/691052431525675048/749335865876021248/1080103960943267911>.
> If you can find a way to make it work without this, or convince us that the
> assessment is wrong / it's worth making an exception, let us know.
>
> Similarly, if you need us to make changes to the internals to allow 3rd
> party crates that need unstable Rust features to implement this externally,
> we're also generally very open to that!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bevyengine/bevy/pull/6161#issuecomment-1448173367>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB3VLDEEHERTHWTCNL26MULWZX32LANCNFSM6AAAAAAQ4QFHE4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

james7132 commented:
Sorry about that, must have been reading an old version of the PR. Really shouldn't be doing this stuff at 4AM, haha.

workingjubilee commented on benches/benches/bevy_ecs/iteration/batched_compute.rs:
```diff
@@ -0,0 +1,238 @@
+// This example is written in a way that benefits from inlined data layout conversion.
+fn aos_to_soa_83(aos_inner: &[Vec3; 8]) -> [__m256; 3] {
+    unsafe {
+        //# SAFETY: Vec3 is repr(C) for x86_64
+        let mx0 = _mm_loadu_ps((aos_inner as *const Vec3 as *const f32).offset(0));
+        let mx1 = _mm_loadu_ps((aos_inner as *const Vec3 as *const f32).offset(4));
+        let mx2 = _mm_loadu_ps((aos_inner as *const Vec3 as *const f32).offset(8));
+        let mx3 = _mm_loadu_ps((aos_inner as *const Vec3 as *const f32).offset(12));
+        let mx4 = _mm_loadu_ps((aos_inner as *const Vec3 as *const f32).offset(16));
+        let mx5 = _mm_loadu_ps((aos_inner as *const Vec3 as *const f32).offset(20));
```
If this is `glam::Vec3`, this is unsound and `repr(C)` is not enough. Vec3, as opposed to Vec3A, has padding bytes. That is, its layout is:

| f32 | f32 | f32 |  ? |
| --- | --- | --- | --- |
| init | init | init | *poison* |

so this is filling the 4th lane with data that should not be read. Vec3A is defined as `__m128` already. You must use a full 16-byte type with *zero* padding, consistently.

InBetweenNames replied:
Thanks for the review!  Indeed, you're right: the C (and C++) standards don't provide any guarantee against arbitrary padding being added between elements or at the end of a struct, even in the case it wouldn't be necessary to guarantee stride alignment in sequential array accesses.  It's perfectly legitimate for a conforming C compiler to lay `glam::Vec3` out in exactly the way you have suggested, including using a trap representation for the padding at the end of the struct.  Even if `glam::Vec3` were defined in terms of a single array `[f32; 3]`, you still can't escape padding at the end as you have indicated, unless you made it `repr(transparent)` anyway.

In C++, the way we'd normally deal with this is with a static assert for `sizeof(Vec3) == 3*sizeof(float)`.  It's a hack, but it works to avoid this problem.  Indeed, on all major x86 ABIs, this holds.  Since static asserts don't really exist in Rust, I'm not sure how to accomplish something equivalent there in this example.

I'd like to mention though, that real x86 C (and C++) compilers would never, ever add padding for the sake of adding it.  Padding is something done on x86 to ensure alignment, and the alignment of `Vec3` is not required to be anything higher than the alignment of `f32`.  It *could* be, and that would be legal in a theoretic sense for C, but no implementation out there -- on x86 -- would ever do it.  For one thing, that would contradict the psABI and break compatibility with other code on x86 systems.  However yes, one could construct a toy C compiler that would do just that, say to anticipate vectorization later on -- although I expect such a transform would be a pessimization a lot more often than not.

So, laying out `Vec3` with 12 bytes of space and aligning to the same alignment of `f32` meets the alignment requirements just fine -- and this is what real implementations -- on x86 -- do.  Not that this excuses anything, mind you, because I agree: the code should be correct from a theoretic sense, or guarded by a compile-time guard to ward off the UB parts (as one would do in C++).

With all this said, as someone that isn't super in the know of the current Rust landscape in dealing with these things, what's the standard practice here?  In C++ I'd stick a static assert on it and be done with it.  How do we make this work in Rust?

workingjubilee replied:
Hmm, it was very late at night when I scanned through things here, and I both mislaid some mental math while reading this and flubbed what I meant to say. My apologies. Let me correct myself: the safety documentation here needs to explicate the actual safety precondition which is the layout match of `8 * 3 / (6 * 4) == 1` and these are unaligned loads, instead of a handwave about `repr(C)`, as if `(n * x) / (m * y) == 1` doesn't hold, **then** you have the possibility of reading the poisoned value.

You needn't particularly worry about toy C compilers, and I wasn't trying to language lawyer you on those. Rust picks specific C compilers to be compatible with (though it gets more vague if e.g. two compilers emit "the same" ABI but actually disagree on fine details). Thus when targeting x86_64-pc-windows-msvc, obviously it emits compatible code with the Windows x64 ABI as emitted by MSVC, etc. etc. down the line.

You can generate a "static assert" in Rust using the pattern
```rust
const _: () = {
    let precondition = {
        // some code that can run in const eval
    };
    if precondition == false {
        panic!("a nicely-written message")
    };
};
```

InBetweenNames replied:
No problem and thanks for the feedback.  I wrote a good amount of this PR myself during some late nights 😅  It sounds like we're converging on a solution.  Would it be acceptable to document this about the expected layout of Vec3 for x86 in the example, and use that pattern to statically assert this is the case for completeness?  With maybe an additional comment about how a `repr(transparent)` type would be preferable for code that needs to be portable?

workingjubilee replied:
Discussing `repr(transparent)` isn't needed but otherwise yes, that sounds fine. My real concern is just about someone retouching this code later without a deep and horrifying knowledge we have of the details of the x86-64 ABI, what `__m128` even is, and C layout, in a way that changes the numbers and causes the the last `_mm_loadu_ps` to read 1 more float than actually exists.

richchurcher commented:
Backlog cleanup: this one has a long history of attempts to get it across the finish line! Closing for now due to inactivity, leaving in place the original tracking issue (#1990) and perhaps it could be adopted from there if still needed.
